stl/inc/inplace_vector

```diff
@@ -0,0 +1,1560 @@
+// inplace_vector standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef _INPLACE_VECTOR_
+#define _INPLACE_VECTOR_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+#include <xmemory>
+
+#include <memory> // TODO: Required for uninitialized range ops without allocator parameter
+
+_STD_BEGIN
+
+template <class _Myvec>
+class _Inplace_vector_const_iterator {
+public:
+    using iterator_concept  = contiguous_iterator_tag;
+    using iterator_category = random_access_iterator_tag;
+    using value_type        = typename _Myvec::value_type;
+    using difference_type   = typename _Myvec::difference_type;
+    using pointer           = typename _Myvec::const_pointer;
+    using reference         = const value_type&;
+
+#if _ITERATOR_DEBUG_LEVEL == 0
+    constexpr _Inplace_vector_const_iterator() noexcept : _Ptr() {}
+
+    constexpr explicit _Inplace_vector_const_iterator(const _Myvec* _Varg, size_t _Off = 0) noexcept
+        : _Ptr(_Varg->data() + _Off) {}
+
+    _NODISCARD constexpr reference operator*() const noexcept {
+        return *_Ptr;
+    }
+
+    _NODISCARD constexpr pointer operator->() const noexcept {
+        return _Ptr;
+    }
+
+    constexpr _Inplace_vector_const_iterator& operator++() noexcept {
+        ++_Ptr;
+        return *this;
+    }
+
+    constexpr _Inplace_vector_const_iterator operator++(int) noexcept {
+        _Inplace_vector_const_iterator _Tmp = *this;
+        ++_Ptr;
+        return _Tmp;
+    }
+
+    constexpr _Inplace_vector_const_iterator& operator--() noexcept {
+        --_Ptr;
+        return *this;
+    }
+
+    constexpr _Inplace_vector_const_iterator operator--(int) noexcept {
+        _Inplace_vector_const_iterator _Tmp = *this;
+        --_Ptr;
+        return _Tmp;
+    }
+
+    constexpr _Inplace_vector_const_iterator& operator+=(const ptrdiff_t _Off) noexcept {
+        _Ptr += _Off;
+        return *this;
+    }
+
+    constexpr _Inplace_vector_const_iterator& operator-=(const ptrdiff_t _Off) noexcept {
+        _Ptr -= _Off;
+        return *this;
+    }
+
+    _NODISCARD constexpr ptrdiff_t operator-(const _Inplace_vector_const_iterator& _Right) const noexcept {
+        return _Ptr - _Right._Ptr;
+    }
+
+    _NODISCARD constexpr reference operator[](const ptrdiff_t _Off) const noexcept {
+        return _Ptr[_Off];
+    }
+
+    _NODISCARD constexpr bool operator==(const _Inplace_vector_const_iterator& _Right) const noexcept {
+        return _Ptr == _Right._Ptr;
+    }

```

Comment:
This is probably buggy, see #4847. How about changing iterator types to `_Inplace_vector_iterator::_Iterator<_IsConst>` and provide comparison operators as hidden friends?

Reply:
```CPP
inplace_vector<expected<any, char>, 1> cont{};
(void) (cont.begin() == cont.end());
```
This works. Is that good enough?